File: app/src/lumen/java/com/cliqz/browser/controlcenter/DashboardFragment.java
package com.cliqz.browser.controlcenter;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.cliqz.browser.R;
import com.cliqz.browser.app.BrowserApp;
import com.cliqz.browser.main.FlavoredActivityComponent;
import com.cliqz.browser.main.Messages;
import com.cliqz.browser.purchases.PurchasesManager;
import com.cliqz.jsengine.Insights;
import com.cliqz.jsengine.ReadableMapUtils;
import com.cliqz.nove.Bus;
import com.cliqz.nove.Subscribe;
import com.facebook.react.bridge.ReadableMap;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import acr.browser.lightning.preference.PreferenceManager;

import static com.cliqz.browser.controlcenter.DashboardItemEntity.VIEW_TYPE_ICON;
import static com.cliqz.browser.controlcenter.DashboardItemEntity.VIEW_TYPE_SHIELD;

/**
 * Copyright © Cliqz 2019
 */
public class DashboardFragment extends ControlCenterFragment {

    private DashboardAdapter mDashboardAdapter;
    private boolean mIsDailyView;
    private final int[] tabsTitleIds = {
            R.string.bond_dashboard_today_title, R.string.bond_dashboard_week_title};

    @Inject
    Bus bus;

    @Inject
    Insights insights;

    @Inject
    PurchasesManager purchasesManager;

    @Inject
    PreferenceManager preferenceManager;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final Bundle arguments = getArguments();
        if (arguments != null) {
            mIsDailyView = arguments.getBoolean(ControlCenterPagerAdapter.IS_TODAY);
        }
        final FlavoredActivityComponent component = getActivity() != null ?
                BrowserApp.getActivityComponent(getActivity()) : null;
        if (component != null) {
            component.inject(this);
        }
        bus.register(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.bond_dashboard_fragment, container,
                false);
        final RecyclerView dashBoardListView = view.findViewById(R.id.dashboard_list_view);
        mDashboardAdapter = new DashboardAdapter(getContext(), bus);
        updateUI();
        dashBoardListView.setAdapter(mDashboardAdapter);
        dashBoardListView.setLayoutManager(new LinearLayoutManager(getContext()));
        changeDashboardState(purchasesManager.isDashboardEnabled() &&
                preferenceManager.isAttrackEnabled() && preferenceManager.getAdBlockEnabled());
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        updateUI();
    }

    public void changeDashboardState(boolean isEnabled) {
        mDashboardAdapter.setIsDashboardEnabled(isEnabled);
    }

    @Override
    public String getTitle(Context context, int pos) {
        return context.getString(tabsTitleIds[pos]);
    }

    @Override
    public void updateUI() {
        if (mDashboardAdapter == null || getView() == null) {
            return;
        }
        if (mIsDailyView) {
            insights.getInsightsData((data) -> getView().post(() -> updateViews(data.getMap("result"))), "day");
        } else {
            insights.getInsightsData((data) -> getView().post(() -> updateViews(data.getMap("result"))), "week");
        }

    }

    @Override
    public void refreshUI() {
        if (getView() == null) {
            return; //return if view is not inflated yet
        }
        mDashboardAdapter.notifyDataSetChanged();
    }

    @Override
    public void refreshUIComponent(boolean optionValue) {
        changeDashboardState(optionValue);
    }

    public void updateViews(ReadableMap data) {
        if (data == null) {
            return;
        }

        final MeasurementWrapper dataSaved;
        final MeasurementWrapper adsBlocked;
        final MeasurementWrapper trackersDetected;
        final MeasurementWrapper pagesVisited;

        if (purchasesManager.isDashboardEnabled() &&
                preferenceManager.getAdBlockEnabled() && preferenceManager.isAttrackEnabled()) {
            dataSaved = ValuesFormatterUtil.formatBytesCount(ReadableMapUtils.getSafeInt(data, "dataSaved"));
            adsBlocked = ValuesFormatterUtil.formatBlockCount(ReadableMapUtils.getSafeInt(data,"adsBlocked"));
            trackersDetected = ValuesFormatterUtil.formatBlockCount(ReadableMapUtils.getSafeInt(data, "trackersDetected"));
            pagesVisited = ValuesFormatterUtil.formatBlockCount(ReadableMapUtils.getSafeInt(data,"pages"));
        } else {
            dataSaved = ValuesFormatterUtil.formatBytesCount(0);
            adsBlocked = ValuesFormatterUtil.formatBlockCount(0);
            trackersDetected = ValuesFormatterUtil.formatBlockCount(0);
            pagesVisited = ValuesFormatterUtil.formatBlockCount(0);
        }

        final List<DashboardItemEntity> dashboardItems = new ArrayList<>();

        dashboardItems.add(new DashboardItemEntity(adsBlocked.getValue(),
                adsBlocked.getUnit() == 0 ? "" : getString(adsBlocked.getUnit()),
                R.drawable.ic_ad_blocking_shiel, getString(R.string.bond_dashboard_ads_blocked_title), -1, VIEW_TYPE_SHIELD));
        dashboardItems.add(new DashboardItemEntity(trackersDetected.getValue(), "",
                R.drawable.ic_eye, getString(R.string.bond_dashboard_tracking_companies_title), -1, VIEW_TYPE_ICON));
        dashboardItems.add(new DashboardItemEntity(dataSaved.getValue(),
                dataSaved.getUnit() == 0 ? "" : getString(dataSaved.getUnit()), R.drawable.ic_ad_blocking_shiel,
                getString(R.string.bond_dashboard_data_saved_title), -1, VIEW_TYPE_SHIELD));
        dashboardItems.add(new DashboardItemEntity(pagesVisited.getValue(), "", R.drawable.ic_anti_phishing_hook,
                getString(R.string.bond_dashboard_phishing_protection_title), -1, VIEW_TYPE_ICON));
        mDashboardAdapter.addItems(dashboardItems);
    }

    @Subscribe
    void onPurchaseCompleted(Messages.PurchaseCompleted purchaseCompleted) {
        if (purchasesManager.getPurchase().isDashboardEnabled()) {
            bus.post(new Messages.EnableAdBlock());
            bus.post(new Messages.EnableAttrack());
            changeDashboardState(true);
        }
    }

}

File: app/src/lumen/java/com/cliqz/browser/controlcenter/DashboardFragment.kt
package com.cliqz.browser.controlcenter

import acr.browser.lightning.preference.PreferenceManager
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.cliqz.browser.R
import com.cliqz.browser.app.BrowserApp
import com.cliqz.browser.extensions.color
import com.cliqz.browser.extensions.setDrawable
import com.cliqz.browser.extensions.tintVectorDrawable
import com.cliqz.browser.main.Messages
import com.cliqz.browser.purchases.PurchasesManager
import com.cliqz.jsengine.Insights
import com.cliqz.jsengine.ReadableMapUtils
import com.cliqz.nove.Bus
import com.cliqz.nove.Subscribe
import com.facebook.react.bridge.ReadableMap
import kotlinx.android.synthetic.lumen.bond_dashboard_fragment.*
import javax.inject.Inject

class DashboardFragment : Fragment() {

    private var isDailyView = false

    @Inject
    internal lateinit var bus: Bus

    @Inject
    internal lateinit var insights: Insights

    @Inject
    internal lateinit var purchasesManager: PurchasesManager

    @Inject
    internal lateinit var preferenceManager: PreferenceManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isDailyView = arguments?.getBoolean(ControlCenterPagerAdapter.IS_TODAY) ?: true
        BrowserApp.getActivityComponent(activity)?.inject(this)
        bus.register(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.bond_dashboard_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        changeDashboardState(purchasesManager.isDashboardEnabled() &&
                preferenceManager.isAttrackEnabled && preferenceManager.adBlockEnabled)
        reset.setOnClickListener {
            if (!preferenceManager.isAttrackEnabled && !preferenceManager.adBlockEnabled) {
                return@setOnClickListener
            }
            android.app.AlertDialog.Builder(it.context)
                    .setTitle(R.string.bond_dashboard_clear_dialog_title)
                    .setMessage(R.string.bond_dashboard_clear_dialog_message)
                    .setPositiveButton(R.string.button_ok) { _, _ ->
                        bus.post(Messages.ClearDashboardData())
                    }
                    .setNegativeButton(R.string.cancel, null)
                    .show()
        }
        updateUI()
    }

    fun getTitle(): String {
        return if (isDailyView) {
            context!!.getString(R.string.bond_dashboard_today_title)
        } else {
            context!!.getString(R.string.bond_dashboard_week_title)
        }
    }

    fun updateUI() {
        val periodType = if (isDailyView) "day" else "week"
        insights.getInsightsData({
            view?.post {
                updateViews(it.getMap("result"))
            }
        }, periodType)
    }

    fun refreshUIComponent(optionValue: Boolean) {
        changeDashboardState(optionValue)
    }

    private fun updateViews(data: ReadableMap?) {
        if (data == null) return

        val dataSaved: MeasurementWrapper
        val adsBlocked: MeasurementWrapper
        val trackersDetected: MeasurementWrapper
        val pagesVisited: MeasurementWrapper

        if (purchasesManager.isDashboardEnabled() &&
                preferenceManager.adBlockEnabled && preferenceManager.isAttrackEnabled) {
            dataSaved = ValuesFormatterUtil.formatBytesCount(ReadableMapUtils.getSafeInt(data, "dataSaved"))
            adsBlocked = ValuesFormatterUtil.formatBlockCount(ReadableMapUtils.getSafeInt(data, "adsBlocked"))
            trackersDetected = ValuesFormatterUtil.formatBlockCount(ReadableMapUtils.getSafeInt(data, "trackersDetected"))
            pagesVisited = ValuesFormatterUtil.formatBlockCount(ReadableMapUtils.getSafeInt(data, "pages"))
        } else {
            dataSaved = ValuesFormatterUtil.formatBytesCount(0)
            adsBlocked = ValuesFormatterUtil.formatBlockCount(0)
            trackersDetected = ValuesFormatterUtil.formatBlockCount(0)
            pagesVisited = ValuesFormatterUtil.formatBlockCount(0)
        }

        ads_blocked.text = adsBlocked.value
        trackers_detected.text = trackersDetected.value
        data_saved.text = dataSaved.value
        phishing_checked.text = pagesVisited.value

        data_saved_icon.setDrawable(when (dataSaved.unit) {
            R.string.bond_dashboard_units_kb -> R.drawable.ic_kb_data_saved_on
            R.string.bond_dashboard_units_mb -> R.drawable.ic_mb_data_saved_on
            R.string.bond_dashboard_units_gb -> R.drawable.ic_gb_data_saved_on
            else -> throw IllegalArgumentException("Wrong unit for 'data saved' data")
        })
    }

    private fun changeDashboardState(isEnabled: Boolean) {
        if (isEnabled) {
            ads_blocked_icon.clearColorFilter()
            trackers_detected_icon.clearColorFilter()
            data_saved_icon.clearColorFilter()
            phishing_checked_icon.clearColorFilter()
            context?.apply {
                ads_blocked.setTextColor(Color.WHITE)
                trackers_detected.setTextColor(Color.WHITE)
                data_saved.setTextColor(Color.WHITE)
                phishing_checked.setTextColor(Color.WHITE)

                ads_blocked_text.setTextColor(Color.WHITE)
                trackers_detected_text.setTextColor(Color.WHITE)
                data_saved_text.setTextColor(Color.WHITE)
                phishing_checked_text.setTextColor(Color.WHITE)

                reset.setTextColor(color(R.color.lumen_color_blue_primary))

                vertical_line.setBackgroundColor(color(R.color.lumen_color_blue_primary_opaque))
                horizontal_line.setBackgroundColor(color(R.color.lumen_color_blue_primary_opaque))
            }
        } else {
            ads_blocked_icon.tintVectorDrawable(R.color.lumen_color_grey_text)
            trackers_detected_icon.tintVectorDrawable(R.color.lumen_color_grey_text)
            data_saved_icon.tintVectorDrawable(R.color.lumen_color_grey_text)
            phishing_checked_icon.tintVectorDrawable(R.color.lumen_color_grey_text)
            context?.apply {
                ads_blocked.setTextColor(color(R.color.lumen_color_grey_text))
                trackers_detected.setTextColor(color(R.color.lumen_color_grey_text))
                data_saved.setTextColor(color(R.color.lumen_color_grey_text))
                phishing_checked.setTextColor(color(R.color.lumen_color_grey_text))

                ads_blocked_text.setTextColor(color(R.color.lumen_color_grey_text))
                trackers_detected_text.setTextColor(color(R.color.lumen_color_grey_text))
                data_saved_text.setTextColor(color(R.color.lumen_color_grey_text))
                phishing_checked_text.setTextColor(color(R.color.lumen_color_grey_text))

                reset.setTextColor(color(R.color.lumen_color_grey_text))

                vertical_line.setBackgroundColor(color(R.color.lumen_color_grey_text))
                horizontal_line.setBackgroundColor(color(R.color.lumen_color_grey_text))
            }
        }
    }

    @Subscribe
    internal fun onPurchaseCompleted(purchaseCompleted: Messages.PurchaseCompleted) {
        if (purchasesManager.purchase.isDashboardEnabled) {
            bus.post(Messages.EnableAdBlock())
            bus.post(Messages.EnableAttrack())
            changeDashboardState(true)
        }
    }

}